src/transaction/buffer.rs

```diff
@@ -176,46 +180,94 @@ impl Buffer {
             .lock()
             .await
             .get(&key)
-            .map(Mutation::get_value)
+            .map(BufferEntry::get_value)
             .unwrap_or(MutationValue::Undetermined)
     }
+
+    fn update_cache(
+        &self,
+        mutations: &mut MutexGuard<BTreeMap<Key, BufferEntry>>,
+        key: Key,
+        value: Option<Value>,
+    ) {
+        match mutations.get(&key) {
+            Some(BufferEntry::ReadLockCached(None)) => {
+                mutations.insert(key, BufferEntry::ReadLockCached(Some(value)));
+            }
+            None => {
+                mutations.insert(key, BufferEntry::Cached(value));
+            }
+            Some(BufferEntry::Cached(v)) | Some(BufferEntry::ReadLockCached(Some(v))) => {
+                assert!(&value == v);
+            }
+            Some(BufferEntry::Put(v)) => {
+                assert!(value.as_ref() == Some(v))
+            }
+            Some(BufferEntry::Del) => {
+                assert!(value.is_none());
+            }
+        }
+    }
 }
 
-// The state of a value in the buffer.
+// The state of a key-value pair in the buffer.
+// It includes two kinds of state:
+//
+// Mutations:
+//   - `Put`
+//   - `Del`
+// Cache of read requests:
+//   - `Cached`, generated by normal read requests
+//   - `ReadLockCached`, generated by lock commands (`lock_keys`, `get_for_update`) and optionally read requests
+//
 #[derive(Debug, Clone)]
-enum Mutation {
+enum BufferEntry {
     // The value has been read from the server. None means there is no entry.
+    // Also means the entry isn't locked.
     Cached(Option<Value>),
+    // Key is locked.
+    //
+    // Cached value:
+    //   - Outer Option: Whether there is cached value
+    //   - Inner Option: Whether the value is empty
+    //   - Note: The cache is not what the lock request reads, but what normal read (`get`) requests read.
+    //
+    // In optimistic transaction:
+    //   The key is locked by `lock_keys`.
+    //   It means letting the server check for conflicts when committing
+    //
+    // In pessimistic transaction:
+    //   The key is locked by `get_for_update` or `batch_get_for_update`
+    ReadLockCached(Option<Option<Value>>),
```

Comment:
I think it's a read lock. It only "writes" a lock but not data. And `select for update` are sometimes called "read lock".

Reply:
So, I think read is just generally ambiguous - IME a write lock prevents writes and a read lock prevents reads. Would it make sense for there to be a `WriteLockCached`? If so, then we should keep the `Read`, but otherwise I don't think we should.

Reply:
Makes sense. 
`CachedLock` is also a little ambiguous. It sounds like the cached content is the lock. 
How about`CachedAndLocked`? Wordy but precise.

Reply:
I think just Lock or Locked is fine - after all they are all cached really.

Reply:
Ah right.